### PR TITLE
Add timeline to db key (YRCORE-4003)

### PR DESCRIPTION
**Pull request by ogrubes**
Sequence is now 4x 32B: LSN1.LSN2.Index.Timeline
Storage DB Key is now ordered as LSN|Index|Timeline|Scope

Clean test run: https://pb.yelpcorp.com/261267

Lots of help from the compiler to catch all these changes :)

### CODE REVIEW

File: common/sequence.go [hunk 1]
```diff
@@ -10,31 +10,34 @@ import (
 	pb "github.com/apigee-labs/transicator/common/proto"
 )
 
-var sequenceRe = regexp.MustCompile("([a-fA-F0-9]+)\\.([a-fA-F0-9]+)\\.([a-fA-F0-9]+)")
+var sequenceRe = regexp.MustCompile("([a-fA-F0-9]+)\\.([a-fA-F0-9]+)\\.([a-fA-F0-9]+)\\.([a-fA-F0-9]+)")
```
Comment (chuckg) [inline common/sequence.go]: Can timeline really be an alpha? Does it get encoded like the halves of the LSN (ie: 32 bit numbers)?

File: common/sequence.go [hunk 2]
```diff
@@ -59,10 +62,15 @@ func ParseSequence(s string) (Sequence, error) {
 	if err != nil {
 		return Sequence{}, fmt.Errorf("Invalid sequence: %s", err)
 	}
+	tl, err := strconv.ParseUint(parsed[4], 16, 32)
```
Comment (chuckg) [inline common/sequence.go]: Do we have a test for this error case?

File: common/sequence.go [hunk 3]
```diff
@@ -93,9 +101,16 @@ func ParseSequenceBytes(b []byte) (Sequence, error) {
 		return Sequence{}, fmt.Errorf("Invalid byte sequence: %s", err)
 	}
 
+	var tl uint32
+	err = binary.Read(buf, networkByteOrder, &tl)
+	if err != nil {
```
Comment (chuckg) [inline common/sequence.go]: Tested?

File: common/sequence_test.go [hunk 4]
```diff
@@ -18,11 +18,13 @@ var _ = Describe("Sequence tests", func() {
...
-		Expect(testSequenceCompare(2, 1, 1, 1)).Should(BeTrue())
-		Expect(testSequenceCompare(1, 1, 1, 2)).Should(BeTrue())
-		Expect(testSequenceCompare(1, 2, 1, 1)).Should(BeTrue())
+		Expect(testSequenceCompare(1, 1, 1, 1, 1, 1)).Should(BeTrue())
+		Expect(testSequenceCompare(1, 2, 1, 1, 1, 1)).Should(BeTrue())
+		Expect(testSequenceCompare(2, 1, 1, 1, 1, 1)).Should(BeTrue())
+		Expect(testSequenceCompare(1, 1, 1, 2, 1, 1)).Should(BeTrue())
+		Expect(testSequenceCompare(1, 1, 2, 1, 1, 1)).Should(BeTrue())
+		Expect(testSequenceCompare(1, 1, 1, 1, 1, 2)).Should(BeTrue())
+		Expect(testSequenceCompare(1, 1, 1, 1, 2, 1)).Should(BeTrue())
```
Comment (chuckg) [inline common/sequence_test.go]: * Do we need to test the reverse? Ie: The none true case?
* Since we're not dealing with timeline remediation; what happens when a timeline switch occurs, and the LSN of timeline 1 is greater than the LSN of timeline 2? This compare would return true inaccurately. I think timeline takes precedence over all other fields as a comparator.

File: storage/convert_test.go [hunk 5]
```diff
@@ -13,101 +13,115 @@ var _ = Describe("Conversion", func() {
...
+
+		s2, cmp := testIndexKeyCompare("foo", "foo", 123, 123, 456, 456, 1, 0)
+		Expect(s2).Should(BeTrue())
+		Expect(cmp).Should(BeEquivalentTo("timeline"))
+
+		s3, cmp := testIndexKeyCompare("foo", "foo", 123, 123, 456, 456, 3, 10)
+		Expect(s3).Should(BeTrue())
+		Expect(cmp).Should(BeEquivalentTo("timeline"))
+	})
 	It("should prioritize lsn, then idx, then scope in key comparison", func() {
```
Comment (chuckg) [inline storage/convert_test.go]: There's a lot of testing going on here; should this be break up into Describe/It blocks? 

```
Describe("LSN sorting")
   It("should prioritize LSN over scope")
   It("should prioritize LSN over index")
```

File: changeserver/replicator.go [hunk 6]
```diff
@@ -49,8 +49,8 @@ func (s *server) handleChange(c *common.Change, firstChange common.Sequence) (co
 			return changeSeq, err
 		}
 		s.db.PutEntryAndMetadata(
-			scope, c.Lsn, c.Index, dataBuf,
-			lastSequenceKey, changeSeq.Bytes())
+			scope, c.Lsn, c.Index, c.Timeline,
```
Comment (chuckg) [inline changeserver/replicator.go]: No subsequent test changes for replicator after this modification?

File: common/sequence_test.go (hunk 4 above)
Comment (kriswehner) [inline common/sequence_test.go]: I don't think timeline ordering is that cut & dried. As long as the keys don't match and clobber each other, and timeline 2 for the same LSN/index is > 1, it's fine. Timeline being after the lsn/index in the key will give us that, and still keep the B+ tree efficient by not putting a huge constant at the front of all the keys.

File: common/proto/logical_output.proto [hunk 7]
```diff
@@ -83,6 +83,8 @@ message Change {
     repeated Column newColumns = 9;
     // Row removed on a delete, or old row for an update.
     repeated Column oldColumns = 10;
+    // Timeline ID to identify WAL. Tertiary sort order.
+    uint32 timeline = 11;
```
Comment (kriswehner) [inline common/proto/logical_output.proto]: I now have a version of the pgoutput plugin that has timeline, so this is solved!

File: common/sequence.go (hunk 1 above)
Comment (kriswehner) [inline common/sequence.go]: Let's ignore this. I'll make a ticket to speed it up.

Comment (kriswehner) [inline common/sequence.go]: It does. everything is hex.

File: common/sequence.go (hunk 2 above)
Comment (ogrubes) [inline common/sequence.go]: ++ added.

File: common/sequence.go (hunk 3 above)
Comment (ogrubes) [inline common/sequence.go]: I don't think I can test this without mocking out binary.Read.
I'm going to add a check for failing when sending in a byte sequence of only < uint64 | uint32 > (no timeline), and call it good.

File: changeserver/checksumapi_test.go [hunk 8]
```diff
@@ -70,16 +70,48 @@ var _ = Describe("Checksum API", func() {
...
+
+		payloadData := make(map[string]string)
+		err := json.Unmarshal(bod, &payloadData)
+		Expect(err).Should(Succeed())
+		Expect(payloadData).Should(HaveKeyWithValue("totalRows", "1"))
+
+		seqValues := [][]string{
+			{"startSequence=0.0.0", "Invalid from value 0.0.0"},
+			{"endSequence=0.0.0", "Invalid to value 0.0.0"},
+			{"startSequence=0.0.0&endSequence=0.0.0", "Invalid from value 0.0.0"},
```
Comment (kriswehner) [inline changeserver/checksumapi_test.go]: Ew, this points out that our error messages stink and reference the old parameter names.

Imma file a bug for this.

File: changeserver/replicator.go (hunk 6 above)
Comment (kriswehner) [inline changeserver/replicator.go]: shouldn't be.
